fxa-rust-client/src/lib.rs

```diff
@@ -90,62 +100,80 @@ impl FirefoxAccount {
         let session_token = hex::decode(credentials.session_token)?;
         let key_fetch_token = hex::decode(credentials.key_fetch_token)?;
         let unwrap_kb = hex::decode(credentials.unwrap_kb)?;
-        let state_data = ReadyForKeysState::new(session_token, key_fetch_token, unwrap_kb);
-        let state = if credentials.verified {
-            EngagedAfterVerified(state_data)
+        let login_state_data = ReadyForKeysState::new(session_token, key_fetch_token, unwrap_kb);
+        let login_state = if credentials.verified {
+            EngagedAfterVerified(login_state_data)
         } else {
-            EngagedBeforeVerified(state_data)
+            EngagedBeforeVerified(login_state_data)
         };
 
         Ok(FirefoxAccount {
-            config,
-            uid: credentials.uid,
-            email: credentials.email,
-            state,
-            oauth_tokens_cache: HashMap::new(),
+            persisted_state: FxAStateV1 {
+                config,
+                uid: credentials.uid,
+                email: credentials.email,
+                login_state,
+                oauth_tokens_cache: HashMap::new(),
+            }
         })
     }
 
     pub fn from_json(data: &str) -> Result<FirefoxAccount> {
-        Ok(serde_json::from_str(data)?)
+        let fxa_state: FxAState = serde_json::from_str(data)?;
+        match fxa_state {
+            FxAState::V1(state) => {
+                Ok(FirefoxAccount {
+                    persisted_state: state
+                })
+            }
+        }
     }
 
     pub fn to_json(&self) -> Result<String> {
-        Ok(serde_json::to_string(self)?)
+        let mut json = serde_json::to_value(&self.persisted_state)?;
+        // Hack: Instead of reconstructing the FxAState enum (and moving self.persisted_state!),
+        // we add the schema_version key manually.
+        if let serde_json::value::Value::Object(mut map) = json {
```

Comment:
This might be more pleasant with a `map_object(|object| ...).ok_or_else(...)` to handle the error case.  You shouldn't be panic!-ing in a function that produces a `Result`, basically.